Mathlib/Order/CompleteLattice/MulticoequalizerDiagram.lean

```diff
@@ -0,0 +1,115 @@
+import Mathlib.CategoryTheory.Limits.Shapes.Multiequalizer
+import Mathlib.CategoryTheory.CommSq
+import Mathlib.Tactic.FinCases
+
+/-!
+# Multicoequalizer diagrams in complete lattices
+
+We introduce the notion of bicartesian square (`Lattice.BicartSq`) in a lattice `T`.
+This consists of elements `x₁`, `x₂`, `x₃` and `x₄` such that `x₂ ⊔ x₃ = x₄` and
+`x₂ ⊓ x₃ = x₁`.
+
+It shall be shown (TODO) that if `T := Set X`, then the image of the
+associated commutative square in the category `Type _` is a bicartesian square
+in a categorical sense (both pushout and pullback).
+
+More generally, if `T` is a complete lattice, `x : T`, `u : ι → T`, `v : ι → ι → T`,
+we introduce a property `MulticoequalizerDiagram x u v` which says that `x` is
+the supremum of `u`, and that for all `i` and `j`, `v i j` is the minimum of `u i` and `u j`.
+Again, when `T := Set X`, we shall show (TOOD) that we obtain a multicoequalizer diagram
+in the category of types.
+
+-/
+
+universe u
+
+open CategoryTheory Limits
+
+namespace Lattice
+
+variable {T : Type u} (x₁ x₂ x₃ x₄ : T) [Lattice T]
+
+/-- A bicartesian square in a lattice consists of elements `x₁`, `x₂`, `x₃` and `x₄`
+such that `x₂ ⊔ x₃ = x₄` and `x₂ ⊓ x₃ = x₁`. -/
+structure BicartSq : Prop where
+  max_eq : x₂ ⊔ x₃ = x₄
+  min_eq : x₂ ⊓ x₃ = x₁
+
+namespace BicartSq
+
+variable {x₁ x₂ x₃ x₄} (sq : BicartSq x₁ x₂ x₃ x₄)
+
+include sq
+
+lemma le₁₂ : x₁ ≤ x₂ := by simpa only [← sq.min_eq] using inf_le_left
+lemma le₁₃ : x₁ ≤ x₃ := by simpa only [← sq.min_eq] using inf_le_right
+lemma le₂₄ : x₂ ≤ x₄ := by simpa only [← sq.max_eq] using le_sup_left
+lemma le₃₄ : x₃ ≤ x₄ := by simpa only [← sq.max_eq] using le_sup_right
+
+/-- The commutative square associated to a bicartesian square in a lattice. -/
+lemma commSq : CommSq (homOfLE sq.le₁₂) (homOfLE sq.le₁₃)
+    (homOfLE sq.le₂₄) (homOfLE sq.le₃₄) := ⟨rfl⟩
+
+end BicartSq
+
+end Lattice
+
+namespace CompleteLattice
+
+variable {T : Type u} [CompleteLattice T] {ι : Type*} (x : T) (u : ι → T) (v : ι → ι → T)
+
+/-- A multicoequalizer diagram in a complete lattice `T` consists of families of elements
+`u : ι → T`, `v : ι → ι → T`, and an element `x : T` such that `x` is the supremum of `u`,
+and for any `i` and `j`, `v i j` is the minimum of `u i` and `u j`. -/
+structure MulticoequalizerDiagram : Prop where
+  iSup_eq : ⨆ (i : ι), u i = x
+  min_eq (i j : ι) : u i ⊓ u j = v i j
+
+namespace MulticoequalizerDiagram
+
+variable {x u v} (d : MulticoequalizerDiagram x u v)
+
+/-- The multispan index in the category associated to the complete lattice `T`
+given by the objects `u i` and the minima `v i j = u i ⊓ u j`,
+when `d : MulticoequalizerDiagram x u v`. -/
+@[simps]
+def multispanIndex : MultispanIndex (.prod ι) T where
+  left := fun ⟨i, j⟩ ↦ v i j
+  right := u
+  fst (_ : ι × ι) := homOfLE (by
+    dsimp
+    rw [← d.min_eq]
+    exact inf_le_left)
+  snd (_ : ι × ι) := homOfLE (by
+    dsimp
+    rw [← d.min_eq]
+    exact inf_le_right)
```

Comment:
If you implement the `grind` annotations above, and add 
```lean
attribute [local grind] MulticoequalizerDiagram MultispanShape.prod_fst MultispanShape.prod_snd
```

somewhere in the section, then both proofs are `by grind` here.